File: pulsar-core/pulsar-tools/pulsar-browser/src/main/java/com/github/kklisura/cdt/protocol/v2023/commands/WebAudio.java
package com.github.kklisura.cdt.protocol.v2023.commands;

/*-
 * #%L
 * cdt-java-client
 * %%
 * Copyright (C) 2018 - 2023 Kenan Klisura
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.github.kklisura.cdt.protocol.v2023.events.webaudio.*;
import com.github.kklisura.cdt.protocol.v2023.support.annotations.EventName;
import com.github.kklisura.cdt.protocol.v2023.support.annotations.Experimental;
import com.github.kklisura.cdt.protocol.v2023.support.annotations.ParamName;
import com.github.kklisura.cdt.protocol.v2023.support.annotations.Returns;
import com.github.kklisura.cdt.protocol.v2023.support.types.EventHandler;
import com.github.kklisura.cdt.protocol.v2023.support.types.EventListener;
import com.github.kklisura.cdt.protocol.v2023.types.webaudio.ContextRealtimeData;

/** This domain allows inspection of Web Audio API. https://webaudio.github.io/web-audio-api/ */
@Experimental
public interface WebAudio {

  /** Enables the WebAudio domain and starts sending context lifetime events. */
  void enable();

  /** Disables the WebAudio domain. */
  void disable();

  /**
   * Fetch the realtime data from the registered contexts.
   *
   * @param contextId
   */
  @Returns("realtimeData")
  ContextRealtimeData getRealtimeData(@ParamName("contextId") String contextId);

  /** Notifies that a new BaseAudioContext has been created. */
  @EventName("contextCreated")
  EventListener onContextCreated(EventHandler<ContextCreated> eventListener);

  /** Notifies that an existing BaseAudioContext will be destroyed. */
  @EventName("contextWillBeDestroyed")
  EventListener onContextWillBeDestroyed(EventHandler<ContextWillBeDestroyed> eventListener);

  /** Notifies that existing BaseAudioContext has changed some properties (id stays the same).. */
  @EventName("contextChanged")
  EventListener onContextChanged(EventHandler<ContextChanged> eventListener);

  /** Notifies that the construction of an AudioListener has finished. */
  @EventName("audioListenerCreated")
  EventListener onAudioListenerCreated(EventHandler<AudioListenerCreated> eventListener);

  /** Notifies that a new AudioListener has been created. */
  @EventName("audioListenerWillBeDestroyed")
  EventListener onAudioListenerWillBeDestroyed(
      EventHandler<AudioListenerWillBeDestroyed> eventListener);

  /** Notifies that a new AudioNode has been created. */
  @EventName("audioNodeCreated")
  EventListener onAudioNodeCreated(EventHandler<AudioNodeCreated> eventListener);

  /** Notifies that an existing AudioNode has been destroyed. */
  @EventName("audioNodeWillBeDestroyed")
  EventListener onAudioNodeWillBeDestroyed(EventHandler<AudioNodeWillBeDestroyed> eventListener);

  /** Notifies that a new AudioParam has been created. */
  @EventName("audioParamCreated")
  EventListener onAudioParamCreated(EventHandler<AudioParamCreated> eventListener);

  /** Notifies that an existing AudioParam has been destroyed. */
  @EventName("audioParamWillBeDestroyed")
  EventListener onAudioParamWillBeDestroyed(EventHandler<AudioParamWillBeDestroyed> eventListener);

  /** Notifies that two AudioNodes are connected. */
  @EventName("nodesConnected")
  EventListener onNodesConnected(EventHandler<NodesConnected> eventListener);

  /**
   * Notifies that AudioNodes are disconnected. The destination can be null, and it means all the
   * outgoing connections from the source are disconnected.
   */
  @EventName("nodesDisconnected")
  EventListener onNodesDisconnected(EventHandler<NodesDisconnected> eventListener);

  /** Notifies that an AudioNode is connected to an AudioParam. */
  @EventName("nodeParamConnected")
  EventListener onNodeParamConnected(EventHandler<NodeParamConnected> eventListener);

  /** Notifies that an AudioNode is disconnected to an AudioParam. */
  @EventName("nodeParamDisconnected")
  EventListener onNodeParamDisconnected(EventHandler<NodeParamDisconnected> eventListener);
}

File: pulsar-core/pulsar-tools/pulsar-browser/src/main/kotlin/ai/platon/cdt/kt/protocol/commands/WebAudio.kt
@file:Suppress("unused")
package ai.platon.cdt.kt.protocol.commands

import ai.platon.cdt.kt.protocol.events.webaudio.AudioListenerCreated
import ai.platon.cdt.kt.protocol.events.webaudio.AudioListenerWillBeDestroyed
import ai.platon.cdt.kt.protocol.events.webaudio.AudioNodeCreated
import ai.platon.cdt.kt.protocol.events.webaudio.AudioNodeWillBeDestroyed
import ai.platon.cdt.kt.protocol.events.webaudio.AudioParamCreated
import ai.platon.cdt.kt.protocol.events.webaudio.AudioParamWillBeDestroyed
import ai.platon.cdt.kt.protocol.events.webaudio.ContextChanged
import ai.platon.cdt.kt.protocol.events.webaudio.ContextCreated
import ai.platon.cdt.kt.protocol.events.webaudio.ContextWillBeDestroyed
import ai.platon.cdt.kt.protocol.events.webaudio.NodeParamConnected
import ai.platon.cdt.kt.protocol.events.webaudio.NodeParamDisconnected
import ai.platon.cdt.kt.protocol.events.webaudio.NodesConnected
import ai.platon.cdt.kt.protocol.events.webaudio.NodesDisconnected
import ai.platon.cdt.kt.protocol.support.annotations.EventName
import ai.platon.cdt.kt.protocol.support.annotations.Experimental
import ai.platon.cdt.kt.protocol.support.annotations.ParamName
import ai.platon.cdt.kt.protocol.support.annotations.Returns
import ai.platon.cdt.kt.protocol.support.types.EventHandler
import ai.platon.cdt.kt.protocol.support.types.EventListener
import ai.platon.cdt.kt.protocol.types.webaudio.ContextRealtimeData
import kotlin.String
import kotlin.Unit

/**
 * This domain allows inspection of Web Audio API.
 * https://webaudio.github.io/web-audio-api/
 */
@Experimental
interface WebAudio {
  /**
   * Enables the WebAudio domain and starts sending context lifetime events.
   */
  suspend fun enable()

  /**
   * Disables the WebAudio domain.
   */
  suspend fun disable()

  /**
   * Fetch the realtime data from the registered contexts.
   * @param contextId
   */
  @Returns("realtimeData")
  suspend fun getRealtimeData(@ParamName("contextId") contextId: String): ContextRealtimeData

  @EventName("contextCreated")
  fun onContextCreated(eventListener: EventHandler<ContextCreated>): EventListener

  @EventName("contextCreated")
  fun onContextCreated(eventListener: suspend (ContextCreated) -> Unit): EventListener

  @EventName("contextWillBeDestroyed")
  fun onContextWillBeDestroyed(eventListener: EventHandler<ContextWillBeDestroyed>): EventListener

  @EventName("contextWillBeDestroyed")
  fun onContextWillBeDestroyed(eventListener: suspend (ContextWillBeDestroyed) -> Unit): EventListener

  @EventName("contextChanged")
  fun onContextChanged(eventListener: EventHandler<ContextChanged>): EventListener

  @EventName("contextChanged")
  fun onContextChanged(eventListener: suspend (ContextChanged) -> Unit): EventListener

  @EventName("audioListenerCreated")
  fun onAudioListenerCreated(eventListener: EventHandler<AudioListenerCreated>): EventListener

  @EventName("audioListenerCreated")
  fun onAudioListenerCreated(eventListener: suspend (AudioListenerCreated) -> Unit): EventListener

  @EventName("audioListenerWillBeDestroyed")
  fun onAudioListenerWillBeDestroyed(eventListener: EventHandler<AudioListenerWillBeDestroyed>): EventListener

  @EventName("audioListenerWillBeDestroyed")
  fun onAudioListenerWillBeDestroyed(eventListener: suspend (AudioListenerWillBeDestroyed) -> Unit): EventListener

  @EventName("audioNodeCreated")
  fun onAudioNodeCreated(eventListener: EventHandler<AudioNodeCreated>): EventListener

  @EventName("audioNodeCreated")
  fun onAudioNodeCreated(eventListener: suspend (AudioNodeCreated) -> Unit): EventListener

  @EventName("audioNodeWillBeDestroyed")
  fun onAudioNodeWillBeDestroyed(eventListener: EventHandler<AudioNodeWillBeDestroyed>): EventListener

  @EventName("audioNodeWillBeDestroyed")
  fun onAudioNodeWillBeDestroyed(eventListener: suspend (AudioNodeWillBeDestroyed) -> Unit): EventListener

  @EventName("audioParamCreated")
  fun onAudioParamCreated(eventListener: EventHandler<AudioParamCreated>): EventListener

  @EventName("audioParamCreated")
  fun onAudioParamCreated(eventListener: suspend (AudioParamCreated) -> Unit): EventListener

  @EventName("audioParamWillBeDestroyed")
  fun onAudioParamWillBeDestroyed(eventListener: EventHandler<AudioParamWillBeDestroyed>): EventListener

  @EventName("audioParamWillBeDestroyed")
  fun onAudioParamWillBeDestroyed(eventListener: suspend (AudioParamWillBeDestroyed) -> Unit): EventListener

  @EventName("nodesConnected")
  fun onNodesConnected(eventListener: EventHandler<NodesConnected>): EventListener

  @EventName("nodesConnected")
  fun onNodesConnected(eventListener: suspend (NodesConnected) -> Unit): EventListener

  @EventName("nodesDisconnected")
  fun onNodesDisconnected(eventListener: EventHandler<NodesDisconnected>): EventListener

  @EventName("nodesDisconnected")
  fun onNodesDisconnected(eventListener: suspend (NodesDisconnected) -> Unit): EventListener

  @EventName("nodeParamConnected")
  fun onNodeParamConnected(eventListener: EventHandler<NodeParamConnected>): EventListener

  @EventName("nodeParamConnected")
  fun onNodeParamConnected(eventListener: suspend (NodeParamConnected) -> Unit): EventListener

  @EventName("nodeParamDisconnected")
  fun onNodeParamDisconnected(eventListener: EventHandler<NodeParamDisconnected>): EventListener

  @EventName("nodeParamDisconnected")
  fun onNodeParamDisconnected(eventListener: suspend (NodeParamDisconnected) -> Unit): EventListener
}
